### PR TITLE
Quite a lot of GUI2 refactoring

### PR DESCRIPTION
**Pull request by Vultraz**
Mostly related to widget initialization.

### CODE REVIEW

File: src/gui/widgets/listbox.cpp [hunk 1]
```diff
@@ -1095,16 +1094,14 @@ widget* builder_horizontal_listbox::build() const
 	list_view* widget = new list_view(
 			true, true, generator_base::horizontal_list, true, list_builder);
 
-	init_control(widget);
+	//init_control(widget);
```
Comment (CelticMinstrel) [inline src/gui/widgets/listbox.cpp]: Delete these rather than commenting them.

Also, I'm wondering where the work of these functions is done now? Presumably I've just missed it.

File: src/gui/core/register_widget.hpp [hunk 2]
```diff
@@ -53,4 +53,10 @@
  *
  * "Calls" REGISTER_WIDGET3(id_definition, id, nullptr)
  */
-#define REGISTER_WIDGET(id) REGISTER_WIDGET3(id##_definition, id, nullptr)
+#define REGISTER_WIDGET(id) REGISTER_WIDGET3(id##_definition, id, nullptr)                                             \
```
Comment (CelticMinstrel) [inline src/gui/core/register_widget.hpp]: Is there some reason for adding this here rather than in `REGISTER_WIDGET3`?

File: src/gui/widgets/listbox.cpp (hunk 1 above)
Comment (Vultraz) [inline src/gui/widgets/listbox.cpp]: The work of init_control is now spread between the widget and styled_widget ctors.

File: src/gui/core/register_widget.hpp (hunk 2 above)
Comment (Vultraz) [inline src/gui/core/register_widget.hpp]: Doesn't build with the declarations of the horizontal and grid listboxes since they have no respective classes.

File: src/gui/dialogs/drop_down_menu.cpp [hunk 3]
```diff
@@ -139,8 +139,7 @@ void drop_down_menu::pre_show(window& window)
 		find_widget<toggle_panel>(&new_row, "panel", false).set_tooltip(entry["tooltip"]);
 
 		if(entry.has_attribute("image")) {
-			image* img = new image();
-			img->set_definition("default");
+			image* img = build_single_widget_instance<image>("image");
```
Comment (Vultraz) [inline src/gui/dialogs/drop_down_menu.cpp]: FTr  the definition is set to "default" if not specified otherwise. As for not making these config arguments, no real reason. I could, I suppose.

File: src/gui/core/window_builder.hpp [hunk 4]
```diff
@@ -116,6 +115,31 @@ builder_widget_ptr build_widget(const config& cfg)
...
+ * allows the widget-specific builder code to be executed.
+ *
+ * @tparam                        The final widget type. The widget pointer will be
+ *                                cast to this.
+ *
+ * @param type                    String ID of the widget type.
+ * @param cfg                     Data config to pass to the widget's builder.
+ */
+template<typename T>
+T* build_single_widget_instance(const std::string& type, const config& cfg = config())
```
Comment (Vultraz) [inline src/gui/core/window_builder.hpp]: Suggestions?

File: src/gui/widgets/scrollbar.hpp [hunk 5]
```diff
@@ -158,10 +159,14 @@ class scrollbar_base : public styled_widget
 	{
 		return step_size_;
 	}
+
 	void set_step_size(const unsigned step_size)
 	{
-		step_size_ = step_size;
-		recalculate();
+		// Step size can't be 0!
+		if(step_size != 0) {
```
Comment (jyrkive) [inline src/gui/widgets/scrollbar.hpp]: If step size of zero isn't allowed, how about throwing an exception instead of just ignoring the call?

File: src/gui/widgets/slider.cpp [hunk 6]
```diff
@@ -93,6 +93,7 @@ void slider::set_value(const int value)
 
 void slider::set_minimum_value(const int minimum_value)
 {
+	std::cerr << "setting min value: " << minimum_value << std::endl;
```
Comment (jyrkive) [inline src/gui/widgets/slider.cpp]: Please remove this line.

File: src/gui/widgets/text_box.cpp [hunk 7]
```diff
@@ -114,6 +114,8 @@ text_box::text_box()
 			&text_box::signal_handler_left_button_up, this, _2, _3));
 	connect_signal<event::LEFT_BUTTON_DOUBLE_CLICK>(std::bind(
 			&text_box::signal_handler_left_button_double_click, this, _2, _3));
+
+	update_offsets();
```
Comment (jyrkive) [inline src/gui/widgets/text_box.cpp]: Have you verified that it's safe to call `update_offsets()` before `set_font_size()` and `set_font_style()`?

File: src/gui/core/window_builder.hpp [hunk 8]
```diff
@@ -100,17 +100,17 @@ register_builder_widget(const std::string& id, widget_builder_func_t functor);
...
  * @param cfg                     The config with the information to
- *                                Instantiate the widget.
+ *                                instantiate the widget.
  *
  * @returns                       A generic widget builder pointer.
  */
-template <class T>
-builder_widget_ptr build_widget(const config& cfg)
+template<typename T>
+builder_widget_ptr make_builder(const config& cfg)
```
Comment (jyrkive) [inline src/gui/core/window_builder.hpp]: What even is the point of this function if it just calls `std::make_shared()`? IMO, it would be clearer to simply call `std::make_shared()` directly.

File: src/gui/dialogs/drop_down_menu.cpp (hunk 3 above)
Comment (CelticMinstrel) [inline src/gui/dialogs/drop_down_menu.cpp]: I don't think it matters that much, just seemed a little weird that only the definition was moved.

File: src/gui/core/window_builder.hpp (hunk 4 above)
Comment (CelticMinstrel) [inline src/gui/core/window_builder.hpp]: Hmm... `build_widget_instance`? Or even just `build_widget`? I was also entertaining the idea of making it an `operator new` with additional arguments (ie, like placement new), not sure if that can work though.

File: src/gui/core/window_builder.hpp (hunk 8 above)
Comment (CelticMinstrel) [inline src/gui/core/window_builder.hpp]: And that would also free up `build_widget` to be the name of the function that _actually_ builds the widget, rather than just creating a builder.

File: src/gui/dialogs/drop_down_menu.cpp (hunk 3 above)
Comment (Vultraz) [inline src/gui/dialogs/drop_down_menu.cpp]: Well, `set_definition` itself was also removed.

File: src/gui/widgets/text_box.cpp (hunk 7 above)
Comment (Vultraz) [inline src/gui/widgets/text_box.cpp]: Should be. Those two functions update the `pango_text` settings, while `update_offsets` handles widget parameters.

**Comment by CelticMinstrel** [comment]
So what about those commented `init_control` calls? They should be removed. Other than that, and the overly long name, this looks good to me.

File: src/gui/widgets/status_label_helper.hpp [hunk 9]
```diff
@@ -62,7 +62,7 @@ default_status_value_getter(T& w)
  */
 template<typename W>
 std::function<void()> bind_status_label(
-		widget& find_in,
+		widget* find_in,
```
Comment (jyrkive) [inline src/gui/widgets/status_label_helper.hpp]: I oppose this whole commit. If a parameter isn't supposed to be nullable, it should be a reference instead of a pointer (barring some rare exceptions).

File: src/gui/widgets/listbox.hpp [hunk 10]
```diff
@@ -325,6 +325,9 @@ class listbox : public scrollbar_container
 	void handle_key_right_arrow(SDL_Keymod modifier, bool& handled) override;
 
 private:
+	/** Helper to update visible area after a key event. */ 
+	void update_visible_area_on_key_event(const bool key_direction_vertical);
```
Comment (jyrkive) [inline src/gui/widgets/listbox.hpp]: Please use an enum to specify the direction instead of a bool.

File: src/gui/widgets/chatbox.cpp [hunk 11]
```diff
@@ -100,9 +100,9 @@ void chatbox::finalize_setup()
 {
 	roomlistbox_ = find_widget<listbox>(this, "room_list", false, true);
 
-	roomlistbox_->set_callback_value_change([this](widget&) {
-		switch_to_window(roomlistbox_->get_selected_row());
-	});
+	// We need to bind a lambda here since switch_to_window is overloaded.
```
Comment (jyrkive) [inline src/gui/widgets/chatbox.cpp]: Please also explain why a lambda alone isn't sufficient. Suggested wording:

    // We need to bind a lambda here since switch_to_window is overloaded.
    // A lambda alone would be more verbose because it'd need to specify all the parameters.

File: src/gui/widgets/toggle_button.cpp [hunk 12]
```diff
@@ -121,7 +121,15 @@ void toggle_button::set_value(const unsigned selected)
 
 	// Check for get_window() is here to prevent the callback from
 	// being called when the initial value is set.
-	if(callback_state_change_ && get_window() != nullptr) {
+	//
+	// TODO: see if this is necessary with the signal handler.
```
Comment (jyrkive) [inline src/gui/widgets/toggle_button.cpp]: As the author of the check: yes, it remains necessary with the signal handler. You may remove the TODO.

File: src/gui/core/event/dispatcher.cpp [hunk 13]
```diff
@@ -198,7 +198,7 @@ bool dispatcher::execute_hotkey(const hotkey::HOTKEY_COMMAND id)
 		return false;
 	}
 
-	return itor->second(*this, id);
+	return itor->second(dynamic_cast<widget&>(*this), id);
```
Comment (jyrkive) [inline src/gui/core/event/dispatcher.cpp]: I wonder, do we even have dispatchers which aren't widgets? Maybe you can outright merge the `dispatcher` and `widget` classes. (Albeit it's understandable if it's outside the scope of this refactoring.)

File: src/gui/widgets/listbox.cpp [hunk 14]
```diff
@@ -532,16 +532,29 @@ void listbox::finalize(builder_grid_const_ptr header,
...
+
 	for(unsigned i = 0, max = std::max(p.get_cols(), p.get_rows()); i < max; ++i) {
-		if(selectable_item* selectable = find_widget<selectable_item>(&p, "sort_" +  std::to_string(i), false, false)) {
-			selectable->set_callback_state_change(std::bind(&listbox::order_by_column, this, i, _1));
+		//
+		// TODO: I had to change this to case to a toggle_button in order to use a signal handler.
+		// Should probably look into a way to make it more general like it was before (used to be
+		// cast to selectable_item).
+		//
+		// vultraz - 8/23/2017
```
Comment (jyrkive) [inline src/gui/widgets/listbox.cpp]: This date format is ambiguous, and not all developers live in countries that use the M/D/Y date format. (Heck, the *only* notable countries which use MDY are USA, Micronesia, and Marshall Islands.)

Please specify the date as [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601): 2017-08-23.

File: src/gui/widgets/listbox.cpp [hunk 15]
```diff
@@ -532,16 +532,29 @@ void listbox::finalize(builder_grid_const_ptr header,
...
+		//
+		if(toggle_button* selectable = find_widget<toggle_button>(&p, "sort_" +  std::to_string(i), false, false)) {
```
Comment (jyrkive) [inline src/gui/widgets/listbox.cpp]: Not your fault, but there are two spaces between the plus sign and the call to `std::to_string()`. Please collapse them to one space.

File: src/scripting/lua_gui2.cpp [hunk 16]
```diff
@@ -780,7 +780,8 @@ int intf_set_dialog_callback(lua_State* L)
 		static dialog_callback_wrapper wrapper;
 		c->connect_click_handler(std::bind(&dialog_callback_wrapper::forward, wrapper, w));
 	} else if(gui2::selectable_item* s = dynamic_cast<gui2::selectable_item*>(w)) {
-		s->set_callback_state_change(&dialog_callback);
+		// TODO: is this the correct way to do this?
```
Comment (jyrkive) [inline src/scripting/lua_gui2.cpp]: Looks correct to me, you can remove the TODO.

File: src/gui/core/event/dispatcher.hpp [hunk 17]
```diff
@@ -882,6 +882,20 @@ inline void disconnect_signal_mouse_left_click(dispatcher& dispatcher,
...
 }
 
+/**
+ * Connects a signal handler for a left mouse button double click.
+ *
+ * I'm not exactly sure why this works in this queue position with toggle
+ * panels, but it does. Will revisit if it becomes an issue later (ie, if
+ * this is used with other widgets and doesn't work).
+ *
+ * vultraz - 8/23/17
```
Comment (jyrkive) [inline src/gui/core/event/dispatcher.hpp]: Please specify the date as ISO 8601.

File: src/gui/core/event/dispatcher.cpp (hunk 13 above)
Comment (CelticMinstrel) [inline src/gui/core/event/dispatcher.cpp]: There might be one? The distributor is not a widget, and may be a dispatcher.

Comment (jyrkive) [inline src/gui/core/event/dispatcher.cpp]: No, the distributor isn't a dispatcher.

File: src/gui/dialogs/loading_screen.cpp [hunk 18]
```diff
@@ -154,49 +159,56 @@ loading_screen* loading_screen::current_load = nullptr;
...
-		// functions (filesystem.cpp config parsing code etc. ) and then use that to end the thread faster.
-
+	/* If the worker thread is running, exit the application to prevent memory corruption.
+	 * TODO: this is still not optimal. The main problem is that this code assumes that this
+	 * happened because the window was closed, which is not necessarily the case (other
+	 * possibilities might be a 'dialog doesn't fit on screen' exception caused by resizing
+	 * the window).
+	 * 
+	 * Another approach might be to add exit points (boost::this_thread::interruption_point())
+	 * to the worker functions (filesystem.cpp config parsing code etc. ) and then use that to
```
Comment (jyrkive) [inline src/gui/dialogs/loading_screen.cpp]: Please remove the space in "etc. )"

File: src/gui/dialogs/end_credits.cpp [hunk 19]
```diff
@@ -73,11 +63,12 @@ void end_credits::pre_show(window& window)
 {
 	window.set_callback_next_draw([this]()
 	{
-		timer_id_ = add_timer(10, std::bind(&end_credits::timer_callback, this), true);
 		// Delay a little before beginning the scrolling
 		last_scroll_ = SDL_GetTicks() + 3000;
 	});
 
+	window.connect_signal<event::DRAW>(std::bind(&end_credits::timer_callback, this), event::dispatcher::front_child);
```
Comment (jyrkive) [inline src/gui/dialogs/end_credits.cpp]: You just added an uninitialized memory read.

The window's callback_next_draw is called **after** drawing the window, and therefore after event::DRAW has been sent. Thus, at the time of the first frame, `last_scroll_` isn't set. Because `timer_callback()` reads `last_scroll_`, it triggers undefined behavior.

Please initialize `last_scroll_` in the constructor initializer list. I suggest setting it to the highest possible uint32_t value because we don't want the first frame to scroll:

    , last_scroll_(std::numeric_limits<uint32_t>::max());

File: src/gui/widgets/listbox.cpp [hunk 20]
```diff
@@ -615,10 +615,19 @@ void listbox::set_active_sorting_option(const order_pair& sort_by, const bool se
 	// TODO: should this be moved to a public header_grid() getter function?
 	grid& header_grid = find_widget<grid>(this, "_header_grid", false);
 
-	selectable_item& wgt = find_widget<selectable_item>(&header_grid, "sort_" + std::to_string(sort_by.first), false);
-	wgt.set_value(static_cast<int>(sort_by.second));
+	selectable_item* w =
+		find_widget<selectable_item>(&header_grid, "sort_" + std::to_string(sort_by.first), false, false);
 
-	order_by_column(sort_by.first, dynamic_cast<widget&>(wgt));
+	if(!w) {
```
Comment (jyrkive) [inline src/gui/widgets/listbox.cpp]: Why did you add this check? Unless this condition can be triggered by UMC, it's better to crash than to silently fail to sort because it alerts us about the problem.

File: src/gui/widgets/listbox.cpp [hunk 21]
```diff
@@ -627,17 +636,19 @@ void listbox::set_active_sorting_option(const order_pair& sort_by, const bool se
 
 const listbox::order_pair listbox::get_active_sorting_option()
 {
-	const auto iter = std::find_if(orders_.begin(), orders_.end(),
-		[](const std::pair<selectable_item*, generator_sort_array>& option) {
-			return option.first != nullptr && option.first->get_value() != SORT_NONE;
+	unsigned int column = 0;
+
+	for(const auto& order : orders_) {
```
Comment (jyrkive) [inline src/gui/widgets/listbox.cpp]: Given that you need the index, it would be better to use a traditional for loop:

    for(unsigned int column = 0; column < orders_.size(); ++column) {
